pulsar-package-management/core/src/main/java/org/apache/pulsar/packages/management/core/impl/PackagesManagementImpl.java

```diff
@@ -139,12 +139,30 @@ public void initialize(PackagesStorage storage) {
             .thenCompose(ignore -> storage.writeAsync(packagePath(packageName), inputStream));
     }
 
+    /***
+     * @see {@link PackagesManagement#delete}.
+     * At this implements, there is a problem:
+     *   Concurrently Another command: `create node` when after "getLogs" and before "deleteLog",
+     *   internal implements is recursive-delete, we could not delete every children-nodes
+     * To resolve this problem, I submit a issue&PR to bookkeeper reporitory:
+     *     https://github.com/apache/bookkeeper/issues/3132
+     *     https://github.com/apache/bookkeeper/pull/3133
+     * TODO After the PR merged, change this implements.
```

Comment:
I think that is a storage issue, not the package management core issue?

Reply:
This pr sovle  problem : "dirty zk node". 

But there has one thing not perfect, the case like this: 

User A create a package: "function://public/default/example@v0.1",  then User A want to delete it, so he exec the command "delete function://public/default/example@v0.1"

At the same time, User B want create package "function://public/default/example@v0.2"

If the process running at this line :
````
if (list == null || list.isEmpty()){ 
    return storage.deleteAsync(packageWithoutVersionPath(packageName)); // running at this line
} else {
    return CompletableFuture.completedFuture(null);
}
````
delete command will delete `example@v0.2`  also. 

So package module need a api for "delete zk-node only empty",  this PR "https://github.com/apache/bookkeeper/pull/3133" provided

Reply:
Don't we need to change the [bk storage implementation](https://github.com/apache/pulsar/blob/85c1ba50cb155111f969403a092affd536d9dcb9/pulsar-package-management/bookkeeper-storage/src/main/java/org/apache/pulsar/packages/management/storage/bookkeeper/BookKeeperPackagesStorage.java#L150)? If the delete path is not empty, we should throw exception from the delete?

Reply:
I still think the TODO should be in the storage API, not the package management API

Reply:
> I still think the TODO should be in the storage API, not the package management API

Hi @mattisonchao , it has been corrected, could you please take a look?